Flutter/casi/lib/casi_user.dart

```diff
@@ -1,134 +1,205 @@
-      {Function(String token, CasiUser user) onSuccess,
-      Function(dynamic err) onError}) {
+  CasiLogin(String clientId, String accessToken, Function? loading,
+      {Function(String token, CasiUser user)? onSuccess,
+      Function(dynamic err)? onError}) {
     this.clientId = clientId;
-
+    this._loading = loading;
     this._onSuccess = onSuccess ?? _onSuccess;
     this._onError = onError ?? _onError;
+    final claimSet = JwtClaim(
+      expiry: new DateTime.now().add(new Duration(minutes: 5)),
+      otherClaims: {
+        'data': {'clientId': clientId}
+      },
+    );
 
-    var builder = new JWTBuilder();
-    var signer = new JWTHmacSha256Signer(accessToken);
-    builder
-      ..expiresAt = new DateTime.now().add(new Duration(minutes: 5))
-      ..setClaim('data', {'clientId': clientId});
-
-    var signedToken = builder.getSignedToken(signer);
-    this.secret = signedToken.toString();
-
-    this._loginURL =
-        "${this._serverUrl}/user/login?serviceURL=${this._serverUrl}/auth/clientVerify?q=${Uri.encodeQueryComponent(this.secret)}";
+    this.secret = issueJwtHS256(claimSet, accessToken);
   }
 
-  CasiLogin.fromToken(String token) {
+  CasiLogin.fromToken(String? token) {
     this._token = token;
   }
 
   Future<void> signIn() async {
-    if (_loginURL == null)
-      throw Exception("No client ID and client secret found");
-    final webview = new FlutterWebviewPlugin();
-    webview.onUrlChanged.listen((url) async {
-      print("URL CHANGED: $url");
-
-      if (url.startsWith("${this._serverUrl}/auth/clientVerify?q=")) {
-        if (_token != null) return;
-        _token = 'mutex';
-        try {
-          _cookies = await webview.getCookies();
-          _token = _cookies['"_rememberme'].toString();
-          _token = _token.trim().substring(0, _token.length - 1);
-          CasiUser user = await fetchUserDetails();
-          this._onSuccess(_token, user);
-        } catch (e) {
-          this._onError("Login Failed");
-        }
-        webview.close();
-      }
-    });
-    _token = null;
-    await webview.launch(
-      _loginURL,
-      ignoreSSLErrors: true,
-      clearCookies: true,
-      clearCache: true,
-      userAgent:
-          'Mozilla/5.0 (Linux; Android 4.0.4; Galaxy Nexus Build/IMM76B) AppleWebKit/535.19 (KHTML, like Gecko) Chrome/18.0.1025.133 Mobile Safari/535.19',
-    );
-  }
+    try {
+      _loading!(true);
+      var requestTokenResponse = await http.post(Uri.parse(_requestURL),
+          body: json.encode({
+            'key1': 'val1',                                        //POST request body to get request token
+            'key2': 'val2',
+          }));
+      String requestToken = json.decode(
+          requestTokenResponse.body)['name'];                     //Key of request token in response
+      print(requestToken);
+      _loginURL =
+          'https://auth.devclub.in';                              //construct login url using request token
```

Comment:
Since these code segments are to be completed once the backend is done implementing them, we can abstract them out as separate functions and call those functions here. This would ensure that the implementation of `signIn` method would not have to change whenever we need to integrate those backend APIs and we could just modify the implementation of those helper functions. So just divide this into two functions - 
1. `getRequestToken() -> string` -- This would make a POST request to the endpoint and return the parsed request token as a string (or maybe store it in one of the private members of the class)
2. `getLoginURL(requestToken: string) -> string` -- This would just be a string substitution function which takes the request token and constructs the login url.